dali/operators/nvcvop/nvcvop.h

```diff
@@ -0,0 +1,183 @@
+ */
+nvcv::Image AsImage(SampleView<GPUBackend> sample, const nvcv::ImageFormat &format);
+
+/**
+ * @brief Wrap a const sample view into an nvcv Image
+ *
+ * @param sample sample view
+ * @param format image format. It needs to match the shape and data type of the sample view
+ */
+nvcv::Image AsImage(ConstSampleView<GPUBackend> sample, const nvcv::ImageFormat &format);
+
+/**
+ * @brief Allocates an image batch using a dynamic scratchpad.
+ * Allocated images have the shape and data type matching the samples in the given TensorList.
+ *
+ * It can be used to allocate a workspace batch that matches a given input batch.
+ */
+void AllocateImagesLike(const TensorList<GPUBackend> &t_list,
+                        kernels::DynamicScratchpad &scratchpad, nvcv::ImageBatchVarShape &output);
+
+/**
+ * @brief Wrap samples from the given TensorList into nvcv Images and push them to an output Image batch
+ */
+void PushImagesToBatch(const TensorList<GPUBackend> &t_list, nvcv::ImageBatchVarShape &batch);
+
+/**
+ * @brief A base class for the CVCUDA operators.
+ * It adds convenience methods to access inputs, outputs and arguments as nvcv types.
+ */
+template <typename BaseOp>
+class NVCVOperator: public BaseOp {
+ public:
+  using BaseOp::spec_;
+  using BaseOp::BaseOp;
+
+  /**
+   * @brief Get an operator argument as an nvcv Tensor.
+   * This method uses the dynamic scratchpad to allocate the (pinned and device) buffers
+   *  used by the output tensor.
+   *
+   * @tparam T data type of the DALI operator argument
+   * @tparam DTYPE expected nvcv Tensor data type
+   * @param arg_shape shape of the DALI operator argument
+   */
+  template <typename T, NVCVDataType DTYPE>
+  nvcv::Tensor AcquireTensorArgument(Workspace &ws, kernels::DynamicScratchpad &scratchpad,
+                                     ArgValue<T, 1> &arg, const TensorShape<> &arg_shape) {
+    auto dtype = nvcv::DataType(DTYPE);
+    int num_samples = ws.GetInputBatchSize(0);
+    int arg_sample_vol = volume(arg_shape);
+    arg.Acquire(spec_, ws, num_samples, arg_shape);
+    T *staging_buffer =
+        scratchpad.AllocatePinned<T>(num_samples * arg_sample_vol, dtype.alignment());
+    for (int s = 0; s < num_samples; ++s) {
+      std::copy(arg[s].data, arg[s].data + arg_sample_vol,
+                staging_buffer + s * arg_sample_vol);
+    }
+    T *device_buffer = scratchpad.AllocateGPU<T>(num_samples * arg_sample_vol, dtype.alignment());
+    MemCopy(device_buffer, staging_buffer, num_samples * arg_sample_vol * sizeof(T), ws.stream());
+    int ndim = arg_shape.sample_dim();
+    auto inner_dim = arg_shape[ndim - 1];
+    DALI_ENFORCE(inner_dim % dtype.numChannels() == 0, "Invalid argument shape.");
+    std::vector<int64_t> shape_data(ndim);
+    shape_data[0] = num_samples;
+    for (int d = 0; d < ndim; ++d) {
+      shape_data[d + 1] = arg_shape[d];
+    }
+    if (inner_dim != dtype.numChannels()) {
+      shape_data.push_back(inner_dim / dtype.numChannels());
+    }
```

Comment:
This overwrites the (incorrectly written) last element of the shape. Is it even used?

Reply:
This shape calculations have to variants. We either remove the inner dimension (e.g, The tensor is {10, 2} of ints, and the dtype is int2 - the resulting tensor is a {10} of int2) or we don't. 

I fixed the bug that was there, moved the shape calculation to a separate function and made it more verbose.